src/plugins/data/public/search/fetch/extract_warnings.ts

```diff
@@ -16,53 +17,38 @@ import { SearchResponseWarning } from '../types';
 export function extractWarnings(rawResponse: estypes.SearchResponse): SearchResponseWarning[] {
   const warnings: SearchResponseWarning[] = [];
 
-  if (rawResponse.timed_out === true) {
+  const isPartial = rawResponse._clusters
+    ? Object.values(
+        (
+          rawResponse._clusters as estypes.ClusterStatistics & {
+            details: Record<string, ClusterDetails>;
+          }
+        ).details
+      ).some((clusterDetails) => clusterDetails.status !== 'successful')
+    : rawResponse.timed_out || rawResponse._shards.failed > 0;
+  if (isPartial) {
     warnings.push({
-      type: 'timed_out',
-      message: i18n.translate('data.search.searchSource.fetch.requestTimedOutNotificationMessage', {
-        defaultMessage: 'Data might be incomplete because your request timed out',
+      type: 'incomplete',
+      message: i18n.translate('data.search.searchSource.fetch.incompleteResultsMessage', {
+        defaultMessage: 'The data might be incomplete or wrong.',
```

Comment:
I understand we're trying to make the messaging more general-purpose to account for multiple issues. But would it make sense to reference shards and partial data within the warning somehow? Perhaps something like: `Index shards returned partial data. Information displayed here may be incomplete or incorrect.`?

Reply:
At this level, the messaging should be high level. It would not be practical to explain why there is incomplete data in this message since there are some many different reasons for incomplete data. More knowledge users (or curious users) can explore the "detailed why" by clicking "view details". In the future, this will open https://github.com/elastic/kibana/issues/166021.